cstore_reader.c

```diff
@@ -147,78 +136,166 @@ CStoreBeginRead(const char *filename, TupleDesc tupleDescriptor,
+	tableFooter = DeserializeTableFooter(footerData);
+
+	pfree(footerData->data);
+	pfree(footerData);
+
+
+	return tableFooter;
+}
+
+
+/*
+ * ReadFooterData reads and returns footer data. The first page contains the
+ * metadata about footer data. The function validates the metadata, reads
+ * footer data from where it is stored and returns it.
+ */
+static StringInfo ReadFooterData(Relation relation)
+{
+	int32 fileLength = 0;
+	BlockNumber totalBlockCount = 0;
+	BlockNumber blockIndex = 0;
+	const int headerLength = sizeof(int32);
+	StringInfo footerData = NULL;
+	Buffer firstBuffer = InvalidBuffer;
+	Page page = NULL;
+	char *pageData = NULL;
+	int pageDataLength = 0;
+	PageHeader pageHeader = NULL;
+	Buffer headerBuffer = InvalidBuffer;
+	Page headerPage = NULL;
+	char *headerData = NULL;
+	uint32 headerDataLength = 0;
+	uint32 startingBlock = 0;
+	uint32 blockCount = 0;
+
+	totalBlockCount = RelationGetNumberOfBlocksInFork(relation, FOOTER_FORKNUM);
+
+	if (totalBlockCount == 0)
+	{
+		return NULL;
+	}
+
+	headerBuffer =  ReadBufferExtended(relation, FOOTER_FORKNUM, 0, RBM_NORMAL,
+									   NULL);
+	LockBuffer(headerBuffer, BUFFER_LOCK_SHARE);
+	headerPage = BufferGetPage(headerBuffer);
+	headerData = PageGetContents(headerPage);
+	pageHeader = (PageHeader) headerPage;
+	headerDataLength =  pageHeader->pd_lower - SizeOfPageHeaderData;
 
-	freeResult = FreeFile(tableFooterFile);
-	if (freeResult != 0)
+	if (!DeserializeTableFooterMetadata(headerData, headerDataLength,
+										&startingBlock, &blockCount))
 	{
-		ereport(ERROR, (errcode_for_file_access(),
-						errmsg("could not close file: %m")));
+		UnlockReleaseBuffer(headerBuffer);
+		return NULL;
 	}
 
-	return tableFooter;
+	firstBuffer = ReadBufferExtended(relation, FOOTER_FORKNUM, startingBlock,
+									 RBM_NORMAL, NULL);
+
+	LockBuffer(firstBuffer, BUFFER_LOCK_SHARE);
+	page = BufferGetPage(firstBuffer);
+	pageHeader = (PageHeader) page;
+	pageData = PageGetContents(page);
+
+	memcpy(&fileLength, pageData, headerLength);
+
+	if (fileLength < CSTORE_POSTSCRIPT_SIZE_LENGTH)
+	{
+		ereport(ERROR, (errmsg("invalid cstore file")));
+	}
+
+	footerData = makeStringInfo();
+
+	pageDataLength = pageHeader->pd_lower - SizeOfPageHeaderData - headerLength;
+
+	appendBinaryStringInfo(footerData, pageData + headerLength, pageDataLength);
+
+	for (blockIndex = 1; blockIndex < blockCount; blockIndex++)
+	{
+		Buffer buffer = ReadBufferExtended(relation, FOOTER_FORKNUM,
+										   blockIndex + startingBlock,
+										   RBM_NORMAL, NULL);
+		LockBuffer(buffer, BUFFER_LOCK_SHARE);
+		page = BufferGetPage(buffer);
+		pageHeader = (PageHeader) page;
+		pageData = PageGetContents(page);
+		pageDataLength = pageHeader->pd_lower - SizeOfPageHeaderData;
+		appendBinaryStringInfo(footerData, pageData, pageDataLength);
+		LockBuffer(buffer, BUFFER_LOCK_UNLOCK);
+		ReleaseBuffer(buffer);
+	}
+
+	UnlockReleaseBuffer(firstBuffer);
+	UnlockReleaseBuffer(headerBuffer);
```

Comment:
That was to prevent any modification during read. Seems like locking only headerBuffer should be enough.